src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -1407,6 +1407,49 @@ void Session::cancel_resumption_delay()
 }
 
 
+void Session::gather_pending_compensating_writes(util::Span<Changeset> changesets,
+                                                 std::vector<ProtocolErrorInfo>* out)
+{
+    if (m_pending_compensating_write_errors.empty()) {
+        return;
+    }
+
+#ifdef REALM_DEBUG
+    REALM_ASSERT_DEBUG(
+        std::is_sorted(m_pending_compensating_write_errors.begin(), m_pending_compensating_write_errors.end(),
+                       [](const ProtocolErrorInfo& lhs, const ProtocolErrorInfo& rhs) {
+                           return lhs.compensating_write_server_version < rhs.compensating_write_server_version;
+                       }));
+#endif
+
+    auto changeset_it = changesets.begin();
+    auto error_it = m_pending_compensating_write_errors.begin();
+    bool found_match = false;
+    while (changeset_it != changesets.end() && error_it != m_pending_compensating_write_errors.end()) {
+        if (error_it->compensating_write_server_version < changeset_it->version) {
+            ++error_it;
+        }
+        else {
+            if (changeset_it->version >= error_it->compensating_write_server_version) {
```

Comment:
is this condition correct? Since it's the else branch, `changeset_it->version` can be at most equal to `error_it->compensating_write_server_version`

Reply:
I'm just implementing std::set_intersection here without actually implementing it. Maybe I screwed it up tho?